### PR TITLE
[Validator] Fix issue 12302 - cache constraints

### PR DESCRIPTION
**Pull request by uwej711**
| Q             | A
| ------------- | ---
| Branch?       | "master" for new features / 2.7, 2.8 or 3.1 for fixes
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #12302
| License       | MIT
| Doc PR        | -

This change allows to still cache constraints even when an un-cachable constraint (i.e. Callback)
is added to a parent class at runtime.

This is achieved by caching the constraints that are loaded for the class in question only
and merging parent and interface constraints after reading from cache.

### CODE REVIEW

File: src/Symfony/Component/Validator/Tests/Mapping/Factory/LazyLoadingMetadataFactoryTest.php [hunk 1]
```diff
@@ -67,8 +67,11 @@ public function testMergeParentConstraints()
             ))),
             new ConstraintA(array('groups' => array(
                 'Default',
+                'EntityParentInterface',
+                'EntityInterfaceB',
                 'Entity',
             ))),
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Mapping/Factory/LazyLoadingMetadataFactoryTest.php]: extra line

File: src/Symfony/Component/Validator/Tests/Mapping/Factory/LazyLoadingMetadataFactoryTest.php [hunk 2]
```diff
@@ -43,10 +44,9 @@ public function testMergeParentConstraints()
         $metadata = $factory->getMetadataFor(self::CLASS_NAME);
 
         $constraints = array(
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Mapping/Factory/LazyLoadingMetadataFactoryTest.php]: extra line

File: src/Symfony/Component/Validator/Tests/Mapping/Factory/LazyLoadingMetadataFactoryTest.php [hunk 3]
```diff
@@ -127,12 +130,43 @@ public function testReadMetadataFromCache()
...
         $this->assertEquals($metadata, $factory->getMetadataFor(self::PARENT_CLASS));
     }
+
+    public function testMetadataCacheWithRuntimeConstraint()
+    {
+        $cache = $this->getMock('Symfony\Component\Validator\Mapping\Cache\CacheInterface');
+        $factory = new LazyLoadingMetadataFactory(new TestLoader(), $cache);
+
+        $cache
+          ->expects($this->any())
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Mapping/Factory/LazyLoadingMetadataFactoryTest.php]: 4 spaces indendation (same below)

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php [hunk 4]
```diff
@@ -109,6 +112,22 @@ public function getMetadataFor($value)
...
+            $this->cache->write($metadata);
+        }
+
+        // Include constraints from the parent class
+        $this->mergeConstraints($metadata);
+
+        return $this->loadedClasses[$class] = $metadata;
+    }
+
+    protected function mergeConstraints(ClassMetadata $metadata)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: can this be made private?

**Comment by nicolas-grekas** [comment]
If this is a bug, it should be submitted again the lowest branch where it applies. Do you know which branch this should be? Can you rebase + change the base branch on github (click on "Edit" near the PR title).

**Comment by uwej711** [comment]
This was already present in 2.3 but as I understand, this version is not maintained any more so I will look at the 2.7 branch.

**Comment by uwej711** [comment]
See https://github.com/symfony/symfony/pull/20793

**Pull request closed without merging**